### ISSUE TITLE
incorrect parsing of Go module or package paths

### ISSUE THREAD

**Issue by ydnar** [dependabot/feedback#188]
We tried out Dependabot Go support and it failed with an odd error, mangling a package URL:

> Dependabot couldn't resolve your project's dependencies as it couldn't access [gcloud-golan](https://github.com/GoogleCloudPlatform/gcloud-golan).

This doesn’t appear in our `go.mod` or `go.sum` files, nor does it match any repo under https://github.com/GoogleCloudPlatform .

**Comment by greysteil** [comment]
I’ve just seen the error report for this and will get it fixed this evening. We only added support for `go.mod` files on Saturday so when we say it’s alpha we really mean it, I’m afraid! That said it’s a big focus for us right now so we should be able to respond to feedback really fast.

**Comment by greysteil** [comment]
OK, slightly embarrassing one - the issue here was a bug in Dependabot's regex for parsing GitHub URLs. Fixed [here](https://github.com/dependabot/dependabot-core/commit/29d0584c86b9c9cc6cb782a3bed43e5ef21b0286) and I'll re-run Dependabot on your repo now, which should close the issue.

Separately, I'm going to add some additional protections so that Dependabot will raise an error on my side, rather than creating an issue on yours, if a similar issue occurs in future.

Thanks for the feedback!

**Comment by ydnar** [comment]
Wow, thanks for the quick turnaround!

**Comment by ydnar** [comment]
@greysteil now it blew up with a new issue, attempting to bump a module dependency to something that doesn’t exist:

```diff
-	github.com/maciekmm/messenger-platform-go-sdk v0.0.0-20180506131304-fe24182d420d
+	github.com/maciekmm/messenger-platform-go-sdk v5.0
```

From what I can tell in your [source](https://github.com/dependabot/dependabot-core/blob/master/lib/dependabot/file_updaters/go/modules/go_mod_updater.rb), Dependabot writes the `go.mod` file directly. You *really* should use the go tools directly (e.g. `go get -u <package>`).

We have been running a nightly job (on CircleCI) that updates our dependencies in this Go project for a couple of years. We recently updated it from `dep` to Go 1.11 modules with the following sequence:

```sh
go get ./...
go list -m -u all | grep \\[ | tee .git/COMMIT_PREPAREDMSG
go get -u -v ./...
go mod tidy -v
git add go.mod go.sum
when-git-dirty create-auto-update-pr "go get -u -v ./..."
```

The `when-git-dirty` and `create-auto-update-pr` are shell commands in the Docker image this job runs in.

**Comment by ydnar** [comment]
We just added Dependabot to an open-source Go repo we manage, ZoneDB: https://github.com/zonedb/zonedb

You’re welcome to use that to debug.

**Comment by greysteil** [comment]
Awesome, thanks!

On the v5 update, it looks like there *is* a v5.0 of that dependency [here](https://github.com/maciekmm/messenger-platform-go-sdk/tree/v5.0), but Dependabot needs to get smarter about how it does major version updates. For now I'm tempted to disable them.

On the file updating, the code is [here](https://github.com/dependabot/dependabot-core/blob/master/helpers/go/updater/main.go). Unfortunately @hmarr, who wrote it, is on holiday for the next few weeks and I'm new to Go, so I'm going to be a little slower at iterating it!

**Comment by ydnar** [comment]
The Facebook package doesn’t appear to be module-aware (go `go.mod` file), and a v5 would necessitate a package path revision e.g. `github.com/maciekmm/messenger-platform-go-sdk/v5`. Since neither are true, Go 1.11 module-aware `go get` can’t resolve that package.

Our repo’s `go.mod` file currently points to `fe24182d420d`, which is the current `HEAD` on `master`.

**Comment by ydnar** [comment]
I think the *right* option is cloning the repo and using the `go` tooling directly to manipulate the `go.mod` file.

The problem here is that `github.com/maciekmm/messenger-platform-go-sdk/v5` doesn’t have a `go.mod` file, and so it’s “v5” isn’t a `v5` that Go 1.11 can understand. According to the Go docs, a major semver release should force a package path change.

Maybe detect if there’s a `go.mod` file present in the repo?

In the meantime, I think your suggestion for #3 (SHA->SHA) will work.

**Comment by ydnar** [comment]
On further reflection, that won’t work. You need to update the `go.sum` file as well, otherwise builds modify it, creating a dirty `git` tree. Our CI fails a build if anything modifies the tree during the test/build process.

**Comment by greysteil** [comment]
You're totally right - I'm waiting on @hmarr to get back from holiday and implement that one!

**Comment by ydnar** [comment]
Great thanks. Should this issue be reopened?

**Comment by greysteil** [comment]
The go.sum work is a separate issue, but wouldn’t hurt to have one open to keep track of it. I’m away from my computer but will open one when I get back.

**Comment by greysteil** [comment]
Created: https://github.com/dependabot/feedback/issues/198

**Comment by hmarr** [comment]
@ydnar a quick update on this - I'm working on our Go modules support at the minute, so it should be quite a bit more solid before long.

The update checking logic is already more sophisticated, so you'll no longer (erroneously) be taken to new major versions - particularly those that don't even follow the Go modules versioning scheme.

Next up will be support for updating the `go.sum` file.

**Comment by hmarr** [comment]
@ydnar if you didn't see it in #198, `go.sum` support just shipped  🎉

We still can't handle major version bumps or run `go mod tidy` as we don't clone the repo, but I'm keen to add that option in the future.
